src/hip_module.cpp

```diff
@@ -563,8 +603,24 @@ hipError_t hipLaunchCooperativeKernelMultiDevice(hipLaunchParams* launchParamsLi
                 launchParamsList[j].stream->criticalData()._av.release_locked_hsa_queue();
 #endif
             }
+            hip_internal::ihipHostFree(tls, mg_sync_ptr);
+            for (int j = 0; j < numDevices; ++j) {
+                hip_internal::ihipHostFree(tls, mg_info_ptr[j]);
```

Comment:
Suppose a few launch successfully, start running, and hit a multigrid barrier.  Then there is a failed launch.  Won't the internal wait in the free wait "forever"?   Should the successful launches be "killed", e.g. by resetting their queues?

Reply:
There is no easy way to reset a queue on the fly without causing further problems. So, a deadlock may occur in your described scenario above. 

To reduce the chance of a deadlock for this particular memory cleanup, I have separated the launch of init_ gws and main kernels into two loops. This way, the GWS gets initialized on all devices before trying to launch the main kernels and that would eliminate the chance of the deadlock here if one of the init_gws launches fails.

The hazard would still remain for the memory clean up later in the code if one of the main kernels fail, but I would say the chance of this failure is low because when we reach to this point to launch the main kernels, all the sanity checks have been done (e.g., valid stream, valid kernel descriptors, successful launch of gws_kernel etc) so the launch hopefully should go in successfully. 

when we start testing this feature, we can design a negative test to particularly test such a scenario.